sql/api/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBaseParser.g4

```diff
@@ -42,6 +42,29 @@ options { tokenVocab = SqlBaseLexer; }
   public boolean double_quoted_identifiers = false;
 }
 
+compoundOrSingleStatement
+    : singleStatement
+    | singleCompound
+    ;
+
+singleCompound
+    : compound SEMICOLON* EOF
+    ;
+
+compound
+    : BEGIN compoundBody END
+    ;
+
+// Last semicolon in body is optional.
+compoundBody
+    : compoundStatement (SEMICOLON+ compoundStatement)* SEMICOLON*
```

Comment:
The + after the SEMICOLON represents "one or more semicolons". Is this intended, do we want to allow more than one semicolon in a row with nothing else in between?

Reply:
Not sure why we did it like this in POC, I've tried in MySql Workbench now and it seems that MySql doesn't allow multiple "empty" semicolons... I'll need to check what is the desired behavior and will update accordingly.

Reply:
Talked with Serge offline. His ref spec says that `;` is mandatory and needs to be present even in the latest statement in the block. This is more restrictive, but the idea is that we can easily make it less restrictive if we need it in the future.

So now, we have only `(compoundStatement SEMICOLON)*` which allows empty `BEGIN END` blocks, but that's also per ref spec.

We can discuss further if we want this changed.

Reply:
Sounds good! Thanks for discussing it!